File: stock_search/models/convex/import_data.py
"""Import the local SQLite data store into the Convex backend."""

from __future__ import annotations

from datetime import UTC, datetime
import os
from pathlib import Path

from dotenv import load_dotenv

from ...api.config import DATA_SQLITE_PATH
from ...sqlite_store import SQLiteStore
from .client import ConvexHttpAdapter
from .convex_schemas import normalize_portfolio_positions, stock_map_to_rows
from .function_names import (
    CONVEX_META_SET,
    CONVEX_PORTFOLIO_SET,
    CONVEX_STOCK_REPLACE_ALL,
)

STATS_GENERATED_AT_KEY = "stats_generated_at"


def run_import_from_local_store(
    *,
    db_path: Path = DATA_SQLITE_PATH,
) -> dict[str, int]:
    """Push the local portfolio and stock data store into Convex."""
    client = ConvexHttpAdapter(
        base_url=os.getenv("CONVEX_URL", ""),
        deploy_key=os.getenv("CONVEX_DEPLOY_KEY", ""),
    )
    store = SQLiteStore(db_path)
    positions = normalize_portfolio_positions(store.load_positions())
    merged_stock_map = store.load_stocks()

    client.mutation(
        CONVEX_PORTFOLIO_SET,
        {
            "key": "default",
            "positions": positions,
        },
    )
    client.mutation(
        CONVEX_STOCK_REPLACE_ALL,
        {
            "rows": stock_map_to_rows(merged_stock_map),
        },
    )
    client.mutation(
        CONVEX_META_SET,
        {
            "key": STATS_GENERATED_AT_KEY,
            "value": datetime.now(tz=UTC).isoformat(),
        },
    )

    return {
        "positions": len(positions),
        "stocks": len(merged_stock_map),
    }


def run_import_from_local_files(
    *,
    db_path: Path = DATA_SQLITE_PATH,
    portfolio_path: Path | None = None,
    stats_path: Path | None = None,
    eval_path: Path | None = None,
) -> dict[str, int]:
    """Deprecated compatibility wrapper for the old JSON import entrypoint."""
    if portfolio_path is not None or stats_path is not None or eval_path is not None:
        raise ValueError("JSON-based import arguments were removed. Use db_path or run_import_from_local_store() with the local SQLite database instead.")
    return run_import_from_local_store(db_path=db_path)


if __name__ == "__main__":
    load_dotenv(".env")
    result = run_import_from_local_store()
    print(result)

File: src/stock-search/models/convex/import-data.ts
/** Import the local SQLite data store into the Convex backend. */

import { config as loadDotenv } from "dotenv";

import { DATA_SQLITE_PATH } from "../../config.js";
import { SQLiteStore } from "../../sqlite-store.js";
import { ConvexHttpClient } from "./client.js";
import {
	normalizePortfolioPositions,
	stockMapToRows,
} from "./convex-schemas.js";
import {
	CONVEX_META_SET,
	CONVEX_PORTFOLIO_SET,
	CONVEX_SECTORS_SET,
	CONVEX_STOCK_REPLACE_ALL,
} from "./function-names.js";

const STATS_GENERATED_AT_KEY = "stats_generated_at";

/** Push the local portfolio and stock data store into Convex. */
export async function runImportFromLocalStore({
	dbPath = DATA_SQLITE_PATH,
}: {
	dbPath?: string;
} = {}): Promise<{ positions: number; stocks: number; sectors: number }> {
	const client = new ConvexHttpClient(
		process.env.CONVEX_URL ?? "",
		process.env.CONVEX_DEPLOY_KEY ?? "",
	);
	const store = new SQLiteStore(dbPath);
	const [positions, mergedStockMap, sectorSnapshot] = await Promise.all([
		store.loadPositions().then(normalizePortfolioPositions),
		store.loadStocks(),
		store.loadSectorSnapshot(),
	]);

	const mutations: Array<Promise<unknown>> = [
		client.mutation(CONVEX_PORTFOLIO_SET, {
			key: "default",
			positions,
		}),
		client.mutation(CONVEX_STOCK_REPLACE_ALL, {
			rows: stockMapToRows(
				Object.fromEntries(
					Object.entries(mergedStockMap).map(([ticker, row]) => [
						ticker,
						{
							indicators: row.indicators,
							evaluation: row.evaluation,
							labels: row.labels,
						},
					]),
				),
			),
		}),
		client.mutation(CONVEX_META_SET, {
			key: STATS_GENERATED_AT_KEY,
			value: new Date().toISOString(),
		}),
	];

	if (sectorSnapshot && sectorSnapshot.sectors.length > 0) {
		mutations.push(
			client.mutation(CONVEX_SECTORS_SET, {
				key: "default",
				sectors: sectorSnapshot.sectors,
				meta: sectorSnapshot.meta,
			}),
		);
	}

	await Promise.all(mutations);

	return {
		positions: positions.length,
		stocks: Object.keys(mergedStockMap).length,
		sectors: sectorSnapshot?.sectors.length ?? 0,
	};
}

/** Deprecated compatibility wrapper for the old JSON import entrypoint. */
export async function runImportFromLocalFiles({
	dbPath = DATA_SQLITE_PATH,
	portfolioPath,
	statsPath,
	evalPath,
}: {
	dbPath?: string;
	portfolioPath?: string | null;
	statsPath?: string | null;
	evalPath?: string | null;
} = {}): Promise<{ positions: number; stocks: number; sectors: number }> {
	if (portfolioPath != null || statsPath != null || evalPath != null) {
		throw new Error(
			"JSON-based import arguments were removed. Use dbPath or runImportFromLocalStore() with the local SQLite database instead.",
		);
	}
	return runImportFromLocalStore({ dbPath });
}

async function main(): Promise<void> {
	loadDotenv({ path: ".env" });
	console.log(await runImportFromLocalStore());
}

if (import.meta.url === `file://${process.argv[1]}`) {
	main().catch((error) => {
		console.error(error);
		process.exit(1);
	});
}
